### PR TITLE
New Adapter: Silverpush

### PR DESCRIPTION
**Pull request by prebid-sp**
Added new server adapter by Silverpush.

Prebid docs request: https://github.com/prebid/prebid.github.io/pull/4748

### CODE REVIEW

File: adapters/silverpush/silverpush.go [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+		}
+	}
+	return
+}
+
+// Builder builds a new instance of the silverpush adapter for the given bidder with the given config.
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &SilverPushAdapter{
```
[the remark this answers is not in the document]
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: This is fixed as required.

File: adapters/silverpush/devicetype.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"regexp"
+	"strings"
+
+	"github.com/prebid/openrtb/v19/openrtb2"
+)
+
+func isMobile(ua string) bool {
+	isValidMobile := regexp.MustCompile(`(ios|ipod|ipad|iphone|android)`)
```
Comment (SyntaxNode) [inline adapters/silverpush/devicetype.go:11]: Please move the regex definition up a level do it's not compiled on every call to these functions.

File: adapters/silverpush/devicetype.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+	return isValidMobile.MatchString(strings.ToLower(ua))
+}
+
+func isCTV(ua string) bool {
+	isCtv := regexp.MustCompile(`(smart[-]?tv|hbbtv|appletv|googletv|hdmi|netcast\.tv|viera|nettv|roku|\bdtv\b|sonydtv|inettvbrowser|\btv\b)`)
+	return isCtv.MatchString(strings.ToLower(ua))
+}
+
+// Check valid Eids
```
Comment (SyntaxNode) [inline adapters/silverpush/devicetype.go:20]: Please follow Go comment conventions, something like:

```
/// isValidEids checks for valid eids.
```

File: adapters/silverpush/devicetype.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+func IsValidEids(eids []openrtb2.EID) bool {
```
Comment (SyntaxNode) [inline adapters/silverpush/devicetype.go:21]: Please start this name with a lower case letter. There is no need to export this function outside of your adapter package.

File: adapters/silverpush/silverpush.go [hunk 5]
```diff
@@ -0,0 +1,346 @@
...
+type SilverPushReqExt struct {
+	PublisherId string  `json:"publisherId"`
+	BidFloor    float64 `json:"bidfloor"`
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{Message: "No impressions in bid request."}}
+	}
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:36]: You do not need to worry about this case. Prebid Server will not call your adapter if there are no impressions. Please remove this check and pull up `ValidateAndProcessRequest(request)` into this method.

File: adapters/silverpush/silverpush.go [hunk 6]
```diff
@@ -0,0 +1,346 @@
...
+
+	return a.ValidateAndProcessRequest(request)
+}
+
+// Validate and Process the request return []*adapters.RequestData and []error
+func (a *adapter) ValidateAndProcessRequest(request *openrtb2.BidRequest) ([]*adapters.RequestData, []error) {
+	imps := request.Imp
+	requests := make([]*adapters.RequestData, 0, len(imps))
+	errors := make([]error, 0, len(imps))
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:45]: Please do not initialize the error array at the start. Let's save on the memory allocations by pushing this to automatically happen later, if needed.

ie:
```
var errors []error
```

File: adapters/silverpush/silverpush.go [hunk 7]
```diff
@@ -0,0 +1,346 @@
...
+
+	for _, imp := range imps {
+		impsByMediaType, err := impressionByMediaType(&imp)
+		if err != nil {
+			errors = append(errors, err)
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:50]: Do you want to exit early here for this kind of error?

File: adapters/silverpush/silverpush.go [hunk 8]
```diff
@@ -0,0 +1,346 @@
...
+		if err := json.Unmarshal(req.User.Ext, &userExtRaw); err != nil {
+			return &errortypes.BadInput{Message: "Invalid user.ext."}
+		}
+		if userExtDataRaw, ok := userExtRaw["data"]; ok {
+			if err := json.Unmarshal(userExtDataRaw, &extUser); err != nil {
+				return &errortypes.BadInput{Message: "Invalid user.ext.data."}
+			}
+			var userCopy = *req.User
+			if IsValidEids(extUser.Eids) {
+				userExt, _ := json.Marshal(
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:140]: Please do not ignore errors.

File: adapters/silverpush/silverpush.go [hunk 9]
```diff
@@ -0,0 +1,346 @@
...
+	return nil
+}
+
+func setExtToRequest(req *openrtb2.BidRequest, publisherID string) {
+
+	record := map[string]string{
+		"bc":          bidderConfig + "_" + bidderVersion,
+		"publisherId": publisherID,
+	}
+	reqExt, _ := json.Marshal(record)
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:158]: Please do not ignore errors.

File: static/bidder-info/silverpush.yaml [hunk 10]
```diff
@@ -0,0 +1,13 @@
+endpoint: "https://prebid.chocolateplatform.co/bidder/?identifier=prebidchoc" 
+maintainer:
+  email: "prebid@silverpush.co"
+gvlVendorID: 42
```
Comment (SyntaxNode) [inline static/bidder-info/silverpush.yaml:4]: This doesn't seem like the correct GVL ID. Please confirm.

File: adapters/silverpush/devicetype.go (hunk 2 above)
Comment (prebid-sp) [reply adapters/silverpush/devicetype.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/silverpush.go (hunk 8 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/silverpush.go (hunk 9 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: static/bidder-info/silverpush.yaml (hunk 10 above)
Comment (prebid-sp) [reply static/bidder-info/silverpush.yaml]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/silverpush.go (hunk 6 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/silverpush.go (hunk 5 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/devicetype.go (hunk 4 above)
Comment (prebid-sp) [reply adapters/silverpush/devicetype.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/devicetype.go (hunk 3 above)
Comment (prebid-sp) [reply adapters/silverpush/devicetype.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)

File: adapters/silverpush/silverpush.go (hunk 7 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go:41]: Yes ! ! Silverpush only support for video and Banner type mediaType. Error is only return when we are not receiving these types. 
`if imp.Banner == nil && imp.Video == nil {
		return nil, &errortypes.BadInput{Message: "Invalid MediaType. SilverPush only supports Banner, Video"}
	}`

File: adapters/silverpush/silverpush.go [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+	}
+
+	return nil
+}
+
+func impressionByMediaType(imp *openrtb2.Imp) ([]openrtb2.Imp, error) {
+
+	if imp.Banner == nil && imp.Video == nil {
+		return nil, &errortypes.BadInput{Message: "Invalid MediaType. SilverPush only supports Banner, Video"}
+	}
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:268]: You do not need this check. `silverpush.yaml` file defines the mediaType that is suuported. Prebid-server core codebase already checks that. Please remove this check.

File: adapters/silverpush/silverpush.go [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+	imps := make([]openrtb2.Imp, 0, 2)
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:269]: I see that you are initialising `imps` to `2` imps however your function will always return just single impression based on your checks below. I think you might wanna change this function definiton as below - 
```

func getImpressionByMediaType(imp *openrtb2.Imp) (openrtb2.Imp) {
              impCopy := *imp
              if imp.Banner != nil {
		impCopy.Video = nil
	     }
             if imp.Video != nil {
		impCopy.Banner = nil
	     }
	return impCopy
```

File: adapters/silverpush/silverpush.go [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}}
+	}
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:294]: This is correct but you could also use the following code - 
```
if adapters.IsResponseStatusCodeNoContent(response) {
		return nil, nil
	}

	if err := adapters.CheckResponseStatusCodeForErrors(response); err != nil {
		return nil, []error{&errortypes.BadInput{
			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
		}}
	}
```

File: adapters/silverpush/silverpush.go [hunk 14]
```diff
@@ -0,0 +1,344 @@
...
+func getMediaTypeForImp(impId string, imps []openrtb2.Imp) (mediaType openrtb_ext.BidType) {
+	mediaType = openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			} else if imp.Video != nil {
+				mediaType = openrtb_ext.BidTypeVideo
+			}
+		}
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:332]: Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.MType field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change.

File: adapters/silverpush/silverpush.go (hunk 7 above)
Comment (SyntaxNode) [reply adapters/silverpush/silverpush.go:41]: Sounds good. Please add a `continue` statement after appending the error to exclude these impressions from the request.

File: adapters/silverpush/silverpush.go [hunk 15]
```diff
@@ -0,0 +1,344 @@
...
+				deviceCopy.DeviceType = 1
+			} else if isCTV(deviceCopy.UA) {
+				deviceCopy.DeviceType = 3
+			} else {
+				deviceCopy.DeviceType = 2
+			}
+		}
+		req.Device = &deviceCopy
+	}
+}
```
Comment (SyntaxNode) [inline adapters/silverpush/silverpush.go:114]: Deeply nested `if` statements are difficult to follow. Please consider inverting the checks in this method and `setUser`. Example:

```
func setDevice(req *openrtb2.BidRequest) {
    if req.Device == nil {
        return
    }

    if len(deviceCopy.UA) == 0 {
        return
    }

    deviceCopy := *req.Device
    deviceCopy.OS = getOS(deviceCopy.UA)

    if isMobile(deviceCopy.UA) {
        deviceCopy.DeviceType = 1
    } else if isCTV(deviceCopy.UA) {
        deviceCopy.DeviceType = 3
    } else {
        deviceCopy.DeviceType = 2
    }

    req.Device = &deviceCopy
}
```

Comment (prebid-sp) [reply adapters/silverpush/silverpush.go:114]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go [hunk 16]
```diff
@@ -0,0 +1,344 @@
...
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, nil
+}
+
+func validateRequest(req *openrtb2.BidRequest) error {
+	imp := &req.Imp[0]
+	var silverPushExt openrtb_ext.ImpExtSilverpush
+	err := setPublisherId(req, imp, &silverPushExt)
```
[the remark this answers is not in the document]
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go (hunk 14 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go (hunk 13 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go:285]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go (hunk 12 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go (hunk 11 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go (hunk 7 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/786baef2e551736d40decb753127e6a90a580b8e)

File: adapters/silverpush/silverpush.go [hunk 17]
```diff
@@ -0,0 +1,333 @@
...
+		impCopy.Banner = nil
+
+	}
+	return impCopy
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:280]: nitpick - you might wanna replace with the following - 
```
if adapters.IsResponseStatusCodeNoContent(response) {
		return nil, nil
}
```

File: adapters/silverpush/silverpush.go [hunk 18]
```diff
@@ -0,0 +1,333 @@
...
+// getMediaTypeForImp figures out which media type this bid is for.
+// SilverPush doesn't support multi-type impressions.
+// If both banner and video exist, take banner as we do not want in-banner video.
+func getMediaTypeForImp(bid openrtb2.Bid) openrtb_ext.BidType {
+	switch bid.MType {
+	case openrtb2.MarkupBanner:
+		return openrtb_ext.BidTypeBanner
+	case openrtb2.MarkupVideo:
+		return openrtb_ext.BidTypeVideo
+
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:320]: nitpick - please remove this blank line.

File: adapters/silverpush/silverpush.go [hunk 19]
```diff
@@ -0,0 +1,333 @@
...
+	setDevice(req)
+
+	if err := setExtToRequest(req, silverPushExt.PublisherId); err != nil {
+		return err
+	}
+	return setImpForAdExchange(imp, &silverPushExt)
+}
+
+func setDevice(req *openrtb2.BidRequest) {
+
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:96]: nitpick - Please remove this blank line.

File: adapters/silverpush/silverpush.go [hunk 20]
```diff
@@ -0,0 +1,333 @@
...
+				userCopy.Ext = userExt
+				req.User = &userCopy
+			}
+		}
+	}
+	return nil
+}
+
+func setExtToRequest(req *openrtb2.BidRequest, publisherID string) error {
+
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:147]: nitpick - Please remove this blank line.

File: adapters/silverpush/silverpush.go [hunk 21]
```diff
@@ -0,0 +1,333 @@
...
+	reqExt, err := json.Marshal(record)
+	if err != nil {
+		return err
+	}
+	req.Ext = reqExt
+	return nil
+}
+
+func setImpForAdExchange(imp *openrtb2.Imp, impExt *openrtb_ext.ImpExtSilverpush) error {
+
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:161]: nitpick - Please remove this blank line.

File: adapters/silverpush/silverpush.go [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+		appCopy.Publisher = &openrtb2.Publisher{ID: impExt.PublisherId}
+		req.App = &appCopy
+
+	}
+
+	return nil
+}
+
+func impressionByMediaType(imp *openrtb2.Imp) openrtb2.Imp {
+
```
Comment (gargcreation1992) [inline adapters/silverpush/silverpush.go:265]: nitpick - Please remove this blank line.

**Review by gargcreation1992 — COMMENTED** [review]
requesting to fix few nitpicks.

File: adapters/silverpush/silverpush.go (hunk 17 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

File: adapters/silverpush/silverpush.go (hunk 18 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

File: adapters/silverpush/silverpush.go (hunk 19 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

File: adapters/silverpush/silverpush.go (hunk 20 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

File: adapters/silverpush/silverpush.go (hunk 21 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

File: adapters/silverpush/silverpush.go (hunk 22 above)
Comment (prebid-sp) [reply adapters/silverpush/silverpush.go]: Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/41daf0668c923b8c72d21af84ad06a0f5fc47a62)

**Review by gargcreation1992 — APPROVED** [review]

**Review by Sonali-More-Xandr — APPROVED** [review]

**Pull request merged**